src/algorithms/highlevel/crosssimilaritymatrix.cpp

```diff
@@ -0,0 +1,136 @@
+"The input feature arrays should be in the shape (x, y), where 'x' is the number of frames.\n\n"
+"An exception is also thrown if either one of the input audio feature arrays are empty or if the output similarity matrix is empty.\n\n"
+"References:\n"
+"[1] Serra, J., Serra, X., & Andrzejak, R. G. (2009). Cross recurrence quantification for cover song identification.New Journal of Physics.\n\n");
+
+
+void CrossSimilarityMatrix::configure() {
+  // configure parameters
+  _frameStackStride = parameter("frameStackStride").toInt();
+  _frameStackSize = parameter("frameStackSize").toInt();
+  _kappa = parameter("kappa").toReal();
+  _toBinary = parameter("toBinary").toBool();
+}
+
+// Construct a 'stacked-frames' feature vector from an input audio feature vector by given 'frameStackSize' and 'frameStackStride'
+std::vector<std::vector<Real> > CrossSimilarityMatrix::stackFrames(std::vector<std::vector<Real> >& frames, int frameStackSize, int frameStackStride) const {
+
+  if (frameStackSize == 1) {
+    return frames;
+  }
+  size_t stopIdx;
+  int increment = frameStackSize * frameStackStride;
+  std::vector<std::vector<Real> > stackedFrames;
+  stackedFrames.reserve(frames.size() - increment);
+  std::vector<Real> stack;
+  for (size_t i=0; i<(frames.size() - increment); i+=frameStackStride) {
+    stopIdx = i + increment;
+    for (size_t startTime=i; startTime<stopIdx; startTime+=frameStackStride) {
+      stack.insert(stack.end(), frames[startTime].begin(), frames[startTime].end());
+    }
+    stackedFrames.push_back(stack);
+    stack.clear();
+  }
+  return stackedFrames;
+}
+
+
+void CrossSimilarityMatrix::compute() {
+  // get inputs and output
+  std::vector<std::vector<Real> > queryFeature = _queryFeature.get();
+  std::vector<std::vector<Real> > referenceFeature = _referenceFeature.get();
+  std::vector<std::vector<Real> >& csm = _csm.get();
+
+  if (queryFeature.empty())
+    throw EssentiaException("CrossSimilarityMatrix: input queryFeature array is empty.");
+  if (referenceFeature.empty())
+    throw EssentiaException("CrossSimilarityMatrix: input referenceFeature array is empty.");
+
+  // construct a new vector by stacking the input features by an specified 'frameStackStride' and 'frameStackSize'
+  std::vector<std::vector<Real> >  queryFeatureStack = stackFrames(queryFeature, _frameStackSize, _frameStackStride);
+  std::vector<std::vector<Real> >  referenceFeatureStack = stackFrames(referenceFeature, _frameStackSize, _frameStackStride);
+
+  std::vector<std::vector<Real> > pdistances;
+  // pairwise euclidean distance
+  pdistances = pairwiseDistance(queryFeatureStack, referenceFeatureStack);
+
+  // check whether to binarize the euclidean cross-similarity matrix using the given threshold kappa
+  if (_toBinary == true) {
+    std::vector<std::vector<Real> > tpDistances = transpose(pdistances);
+    size_t queryRows = pdistances.size();
```

Comment:
It's hard to know what ```queryRows``` and ```referenceRows``` represent from the code. Let's make it even more code-readable. 

Can we call  ```queryRows``` something like a ```queryFeatureSize```? (same for reference)